src/metabase/lib/metadata/cached_provider.cljc

```diff
@@ -51,18 +51,19 @@
 (defn- metadatas [cache uncached-provider metadata-type ids]
   (when (seq ids)
     (log/tracef "Getting %s metadata with IDs %s" metadata-type (pr-str (sort ids)))
-    (let [existing-ids (set (keys (get @cache metadata-type)))
-          missing-ids  (set/difference (set ids) existing-ids)]
-      (log/tracef "Already fetched %s: %s" metadata-type (pr-str (sort (set/intersection (set ids) existing-ids))))
-      (when (seq missing-ids)
-        (log/tracef "Need to fetch %s: %s" metadata-type (pr-str (sort missing-ids)))
-        ;; TODO -- we should probably store `::nil` markers for things we tried to fetch that didn't exist
-        (doseq [instance (lib.metadata.protocols/metadatas uncached-provider metadata-type missing-ids)]
-          (store-in-cache! cache [metadata-type (:id instance)] instance))))
+    (let [metadata-cache (get @cache metadata-type)]
+      (when-not (every? #(contains? metadata-cache %) ids)
+        (let [existing-ids (set (keys metadata-cache))
+              missing-ids  (set/difference (set ids) existing-ids)]
+          (log/tracef "Already fetched %s: %s" metadata-type (pr-str (sort (set/intersection (set ids) existing-ids))))
+          (when (seq missing-ids)
+            (log/tracef "Need to fetch %s: %s" metadata-type (pr-str (sort missing-ids)))
+            ;; TODO -- we should probably store `::nil` markers for things we tried to fetch that didn't exist
+            (doseq [instance (lib.metadata.protocols/metadatas uncached-provider metadata-type missing-ids)]
+              (store-in-cache! cache [metadata-type (:id instance)] instance))))))
```

Comment:
We're skipping this side-effect, and the "Already fetched" log line. Are you sure this is OK?

Reply:
Whoops, nevermind, eyesight failure.